File: problems/208/solution.go
package main
const R = 26

type Trie struct {
    links [R]*Trie
    isEnd bool
}


/** Initialize your data structure here. */
func Constructor() Trie {
    return Trie{}
}


/** Inserts a word into the trie. */
func (this *Trie) Insert(word string)  {
    node := this
    
    for _, v := range word {
        if !node.containsKey(v) {
            node.put(v, &Trie{})
        }
        
        node = node.get(v)
    }
    
    node.isEnd = true
}


/** Returns if the word is in the trie. */
func (this *Trie) Search(word string) bool {
    node := this.searchPrefix(word)
    
    return node != nil && node.isEnd
}


/** Returns if there is any word in the trie that starts with the given prefix. */
func (this *Trie) StartsWith(prefix string) bool {
    node := this.searchPrefix(prefix)
    
    return node != nil
}

func (this *Trie) containsKey(ch rune) bool {
    return this.links[ch - 'a'] != nil
}

func (this *Trie) get(ch rune) *Trie {
    return this.links[ch - 'a']
}

func (this *Trie) put(ch rune, t *Trie) {
    this.links[ch - 'a'] = t
}

func (this *Trie) searchPrefix(prefix string) *Trie {
    node := this
    
    for _, v := range prefix {
        if !node.containsKey(v) {
            return nil
        }
        
        node = node.get(v)
    }
    
    return node
}


/**
 * Your Trie object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Insert(word);
 * param_2 := obj.Search(word);
 * param_3 := obj.StartsWith(prefix);
 */

File: problems/208/solution.py
class Trie:
    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.childs = {}
        self.isEnd = False
        

    def insert(self, word: str) -> None:
        """
        Inserts a word into the trie.
        """
        node = self
        
        for v in word:
            ch = ord(v) - 97
            
            if ch not in node.childs:
                node.childs[ch] = Trie()
            
            node = node.childs[ch]
        
        node.isEnd = True

    def search(self, word: str) -> bool:
        """
        Returns if the word is in the trie.
        """
        node = self
        
        for v in word:
            ch = ord(v) - 97
            
            if ch not in node.childs:
                return False
            
            node = node.childs[ch]
            
        return node and node.isEnd

    def startsWith(self, prefix: str) -> bool:
        """
        Returns if there is any word in the trie that starts with the given prefix.
        """
        node = self
        
        for v in prefix:
            ch = ord(v) - 97
            
            if ch not in node.childs:
                return False
            
            node = node.childs[ch]
        
        return True


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)